### ISSUE TITLE
feat(server): /v1/rerank with sequence-classifier, Qwen3 generative and Qwen3-VL multimodal rerankers

### ISSUE THREAD

**Issue by inureyes** [lablup/mlxcel#1356]
## Summary

mlxcel cannot score query/document relevance. This issue adds `POST /v1/rerank` (Cohere and Jina compatible request and response), a `--reranker-model` flag with its own worker thread, and three reranker kinds: one-label sequence classifiers on the BERT, XLM-RoBERTa and ModernBERT encoders (`score = sigmoid(logit)`), the Qwen3 generative reranker (`score = sigmoid(logit("yes") - logit("no"))` at the last prompt position) and the Qwen3-VL multimodal reranker (same scoring, documents and queries may carry images). Sequence classifiers are detected from `architectures[0]` ending in `ForSequenceClassification` and can be served from `-m` directly.

## Current behavior

- `src/server/app.rs:118-167` has no rerank route; `src/server/types/` has no rerank request type.
- `src/models/detection.rs:124-199` never reads `architectures`, so `XLMRobertaForSequenceClassification` is `Unsupported model type: xlm-roberta` and `Qwen3-Reranker` (`model_type: qwen3`, `Qwen3ForCausalLM`) loads as a chat model with no way to read yes/no logits for a batch of prompts.
- #1321 and #1332 deliver `BertSequenceClassifier::logits` and `ModernBertSequenceClassifier::logits` (`[B, num_labels]`) without any HTTP wiring; #1353 delivers `encode_pairs` (pair tokenization with `token_type_ids`) and the padding-mask builders.
- `Qwen3Model::forward` (`src/models/qwen3.rs:676`) accepts an explicit mask and returns `[B, L, vocab]` logits, which is all the generative path needs; `crate::models::Qwen3VLModel::forward_for_sequence` (`src/models/qwen3_vl.rs:841`) does the same with M-RoPE and DeepStack.

### HTTP surface

`POST /v1/rerank` (and `/rerank`):
```json
{
  "model": "optional; must match the served reranker id",
  "query": "string | {\"text\": ..., \"image\": url | \"image_url\": url}",
  "documents": ["string | {text, image | image_url}", ...],   // at least one
  "top_n": 3,                  // optional, >= 1
  "return_documents": false,
  "instruction": "optional, generative rerankers only"
}
```
Response:
```json
{
  "model": "<served id>",
  "results": [{"index": 1, "relevance_score": 0.98, "document": <request item, only when return_documents>}, ...],
  "usage": {"prompt_tokens": 312, "total_tokens": 312}
}
```
`results` is sorted by `relevance_score` descending, ties by ascending `index`, truncated to `top_n`. `prompt_tokens` is the number of real tokens over every scored pair. Errors: empty or whitespace-only `query`/document text, `documents` empty, `top_n < 1`, an image item for a text-only reranker, `instruction` given to a sequence classifier, and `model` mismatch are `400`; no reranker loaded is `501`; queue full `503`; timeout `504`.

### Reranker kinds and detection

| kind | selected when | checkpoint |
|---|---|---|
| `SequenceClassifier` | `architectures[0]` ends with `ForSequenceClassification` and `model_type` in {`bert`, `xlm-roberta`, `modernbert`}; `num_labels` (or `len(id2label)`) must be 1, otherwise a load error "reranker sequence classifiers must expose exactly one output label" | `BAAI/bge-reranker-v2-m3`, `cross-encoder/ms-marco-MiniLM-L6-v2`, `Alibaba-NLP/gte-reranker-modernbert-base` |
| `GenerativeText` | `--reranker-model` points at a `model_type: qwen3` causal checkpoint | `mlx-community/Qwen3-Reranker-0.6B-4bit` |
| `GenerativeVl` | `--reranker-model` points at a `model_type: qwen3_vl` checkpoint (its `modules.json` contains `1_LogitScore`, not a Pooling module) | `Qwen/Qwen3-VL-Reranker-2B` |

Any other `model_type` or a `DebertaV2ForSequenceClassification` is rejected with "Unsupported reranker model type". `-m <checkpoint>` auto-detects only the `SequenceClassifier` kind (a new `ModelType::SequenceClassifier` with `ModelKind::Reranker`, handled like `Embedding` in `load_model`); generative rerankers are indistinguishable from chat models and are loaded through `--reranker-model`. When `--reranker-model` names the same path as `-m`, the checkpoint is loaded once, on the reranker worker, and chat stays unloaded.

### Sequence classifier scoring

```
for each document d (batches of --rerank-batch-size, default 8):
    ids, mask, type_ids = encode_pairs(tokenizer, [(query, d)], max_length)   # [CLS] q [SEP] d [SEP]; XLM-R: <s> q </s></s> d </s>
    logits = classifier.logits(ids, mask, type_ids)   # [B, 1]
    score = sigmoid(logits[:, 0])
```
`max_length = min(8192, config.max_position_embeddings (minus `pad_token_id + 1` for XLM-R), tokenizer model_max_length when < 1e6)`; truncation drops tokens from the longer side first (longest-first) and never drops the query's special tokens.

### Qwen3 generative scoring

Prompt per pair (exact bytes):
```
PREFIX  = "<|im_start|>system\nJudge whether the Document meets the requirements based on the Query and the Instruct provided. Note that the answer can only be \"yes\" or \"no\".<|im_end|>\n<|im_start|>user\n"
CONTENT = "<Instruct>: {instruction}\n<Query>: {query}\n<Document>: {document}"
SUFFIX  = "<|im_end|>\n<|im_start|>assistant\n<think>\n\n</think>\n\n"
tokens  = encode(PREFIX, no special) + encode(CONTENT, no special) + encode(SUFFIX, no special)
```
Default instruction: `Given a web search query, retrieve relevant passages that answer the query`. `CONTENT` is truncated so that the total length is at most 8192 (the suffix is always kept). Batches are left-padded with `pad_token_id` (`<|endoftext|>`, 151643) so every row's last prompt token is at position `L - 1`; the mask is `create_causal_padding_mask(mask, 0)` (`[B,1,L,L]`, with the fully-padded-row rescue). `logits = Qwen3Model::forward(ids, fresh caches, Some(mask))[:, L - 1, :]`; `score = sigmoid(logits[yes_id] - logits[no_id])` where `yes_id`/`no_id` are resolved at load by encoding `yes` and `no` without special tokens (9693 and 2152 for the Qwen3 tokenizer) and the load fails if either is not a single token. Batch size default 8.

### Qwen3-VL generative scoring

Messages rendered with the checkpoint's `additional_chat_templates/reranker.jinja` (`add_generation_prompt: true`):
```
[{"role": "system",   "content": [{"type": "text", "text": instruction}]},        # omitted -> template default "Given a search query, retrieve relevant candidates that answer the query."
 {"role": "query",    "content": [{"type": "image"}?, {"type": "text", "text": query.text}?]},
 {"role": "document", "content": [{"type": "image"}?, {"type": "text", "text": document.text}?]}]
```
which yields `<|im_start|>system\nJudge whether ... "yes" or "no".<|im_end|>\n<|im_start|>user\n<Instruct>: {instruction}<Query>:{query}\n<Document>:{document}<|im_end|>\n<|im_start|>assistant\n`, with each image rendered as `<|vision_start|><|image_pad|><|vision_end|>` and expanded by the Qwen3-VL processor. `1_LogitScore/config.json` supplies `true_token_id` (9693) and `false_token_id` (2152). Images are merged through `Qwen3VLModel::get_input_embeddings`, rows are left-padded, the mask is `create_causal_padding_mask`, logits come from `forward_for_sequence(ids, Some(embeds), caches, Some(mask))[:, L - 1, :]`, `score = sigmoid(logit[true] - logit[false])`. Batch size default 2 (`--rerank-batch-size` overrides).

### Flags

`--reranker-model <path>` (env `MLXCEL_RERANKER_MODEL`, `LLAMA_ARG_RERANKER_MODEL`), `--rerank-batch-size N` (env `MLXCEL_RERANK_BATCH_SIZE`), queue depth and timeout shared with the embedding worker flags. `mlxcel rerank -m <path> -q "query" -d "doc1" -d "doc2" [--image <file>] [--instruction ...]` prints scores in document order and the ranking, the offline validation tool.

## Implementation plan

1. `src/rerank/mod.rs` (new): `RerankerKind`, `RerankItem { text: Option<String>, image: Option<ImageInput> }`, `trait Reranker { fn kind(&self) -> RerankerKind; fn score(&self, query: &RerankItem, docs: &[RerankItem], instruction: Option<&str>) -> Result<(Vec<f32>, usize /* prompt tokens */)>; fn supports_images(&self) -> bool; }`, `detect_reranker_kind(config) -> Result<RerankerKind>`, `num_labels(config)`.
2. `src/rerank/sequence_classifier.rs`: wraps `BertSequenceClassifier` / `ModernBertSequenceClassifier` behind one enum, uses `encode_pairs`, batches, `mlxcel_core::sigmoid`.
3. `src/rerank/qwen3_generative.rs`: `Qwen3Reranker { model: Qwen3Model, tokenizer, yes_id, no_id, prefix_ids, suffix_ids, max_length: 8192 }` with the prompt builder above, left padding (`encode_batch(..., padding_side = Left)`, a new option on the infra helper), `create_causal_padding_mask`, last-position logits via `slice`, `subtract`, `sigmoid`, `eval`, `to_vec_f32`.
4. `src/rerank/qwen3_vl_generative.rs`: `Qwen3VlReranker { vlm: crate::vision::Qwen3VLModel, template: ChatTemplateProcessor (loaded from additional_chat_templates/reranker.jinja), true_id, false_id }`; renders messages, runs the processor for images, scores as above. The `role: query` / `role: document` messages are passed to the template verbatim (the template owns the rendering).
5. `src/models/detection.rs`: `is_sequence_classifier(config)` -> `ModelType::SequenceClassifier` (kind `Reranker`) before the embedding check; `load_model` bails for `ModelKind::Reranker`. `src/rerank/loader.rs`: `load_reranker(path) -> Box<dyn Reranker>` by kind.
6. `src/server/rerank_model.rs` + `src/server/rerank_worker.rs`: `RerankModelProvider` trait and a worker built on the same bounded-queue design as the embedding worker (share the generic worker if #1353 extracted one). `AppState.rerank_model: Option<Arc<dyn RerankModelProvider>>`, `with_rerank_model`; startup wiring next to the embedding slot; flags in `src/bin/mlx_server.rs`, `src/server/cli_input.rs`, `src/server/config.rs`.
7. `src/server/types/rerank.rs` (serde untagged `RerankInput`), `src/server/routes/rerank.rs` (`rerank` handler: normalize items, validate, `spawn_blocking`, sort, `top_n`, `return_documents`, metrics), mount `/v1/rerank` and `/rerank` in `src/server/app.rs`; `/v1/models` lists the reranker.
8. `src/commands/rerank.rs` + `Commands::Rerank`; `docs/embeddings.md` rerank section (schema, kinds, prompt format, flags); `docs/supported-models.md` reranker table.

## Validation

(a) Unit tests:
- `src/rerank/mod_tests.rs`: `detects_sequence_classifier_for_bert_modernbert_xlm_roberta`, `rejects_multi_label_classifier`, `rejects_unsupported_family` (`deberta-v2`), `rejects_embedding_checkpoint_as_reranker` (`BertModel`), `logit_score_modules_json_is_generative_vl_not_embedding`.
- `src/rerank/qwen3_generative_tests.rs`: `prompt_bytes_match_reference_recipe` (exact string equality of PREFIX/CONTENT/SUFFIX), `content_truncation_keeps_suffix`, `left_padding_puts_last_token_at_l_minus_1`, `yes_no_ids_must_be_single_tokens`, `sigmoid_of_logit_difference` on a stub model returning fixed logits.
- `src/rerank/sequence_classifier_tests.rs`: `pairs_are_encoded_with_type_ids_for_bert`, `longest_first_truncation_keeps_query`.
- `src/server/routes/rerank_tests.rs` (stub provider): `results_sorted_desc_ties_by_index`, `top_n_truncates`, `return_documents_echoes_items`, `empty_documents_is_400`, `image_item_on_text_reranker_is_400`, `instruction_on_sequence_classifier_is_400`, `no_reranker_is_501`.

(b) Real checkpoints:
```bash
./target/release/mlxcel download BAAI/bge-reranker-v2-m3
./target/release/mlxcel rerank -m models/bge-reranker-v2-m3 -q "what is panda?" -d "hi" \
  -d "The giant panda (Ailuropoda melanoleuca), sometimes called a panda bear or simply panda, is a bear species endemic to China."
```
Acceptance: scores `[0.0003, 0.9948]` within 2e-2 (the normalized values published in the checkpoint's model card), i.e. raw logits near `-8.19` and `5.26`.

```bash
./target/release/mlxcel download mlx-community/Qwen3-Reranker-0.6B-4bit
./target/release/mlxcel-server -m models/Qwen3-4B-4bit --reranker-model models/Qwen3-Reranker-0.6B-4bit --port 8080
curl -s localhost:8080/v1/rerank -H 'Content-Type: application/json' -d '{
  "query": "What is the capital of China?",
  "documents": ["The capital of China is Beijing.", "Gravity is a force that attracts two bodies towards each other.", "Berlin is in Germany."],
  "top_n": 2, "return_documents": true}'
```
Acceptance: `results[0].index == 0` with `relevance_score > 0.9`, the other two below 0.2, `results` has two entries, and `/v1/chat/completions` still answers from the 4B model in the same process. `cross-encoder/ms-marco-MiniLM-L6-v2` and `Alibaba-NLP/gte-reranker-modernbert-base` must rank the Beijing document first for the same request through `-m` alone.

```bash
./target/release/mlxcel download Qwen/Qwen3-VL-Reranker-2B
./target/release/mlxcel rerank -m models/Qwen3-VL-Reranker-2B -q "a chart of quarterly revenue" --image /path/to/revenue-chart.png --image /path/to/cat.jpg
```
Acceptance: the chart image scores above the cat image by at least 0.3, finite scores for a batch of two image documents (exercises left padding with images), and text-only pairs through the same model rank the Beijing document first.

## Acceptance criteria

- [x] `POST /v1/rerank` and `/rerank` implement the schema above with sorting, `top_n`, `return_documents`, usage and the error codes
- [x] sequence classifiers (BERT, XLM-RoBERTa, ModernBERT, one label) load from `-m` or `--reranker-model` and score with `sigmoid(logit)`
- [x] the Qwen3 generative reranker uses the exact prefix/content/suffix recipe, left padding, and `sigmoid(logit(yes) - logit(no))`
- [x] the Qwen3-VL reranker renders the checkpoint's `reranker.jinja` with image items and reads `1_LogitScore/config.json`
- [x] `--reranker-model`, `--rerank-batch-size` and `mlxcel rerank` exist and are documented
- [x] bge-reranker-v2-m3 and Qwen3-Reranker-0.6B-4bit pass the gates above
- [x] docs/supported-models.md updated
- [x] detection table in src/models/detection.rs updated with a test
- [ ] cargo test --workspace --profile test-fast --features metal,accelerate passes (not runnable on the Linux/CUDA validation host; the CUDA equivalents were run per module and pass)
- [x] cargo clippy --features cuda --lib --bins --tests -- -D warnings and cargo fmt --all -- --check pass (the validation host is the Linux/CUDA runner, where a workspace-wide all-targets clippy is not runnable)

## Out of scope

- Video documents for the Qwen3-VL reranker.
- Serving a generative reranker from the chat model worker (shared weights); it always runs on its own worker.
- Multi-label classifiers and listwise rerankers.

## Dependencies

- Blocked by #1353
- Blocked by #1321
- Blocked by #1332

- Part of epic #1348
